chctools/horn_hacking_lemmas.py

```diff
@@ -0,0 +1,710 @@
+            return
+
+        rels = list()
+        for r in self._rules:
+            rels.extend(r.used_rels())
+        for q in self._queries:
+            rels.extend(q.used_rels())
+        self._rels_set = frozenset(rels)
+        self._sealed = True
+
+        for rel in self._rels_set:
+            self._rels[str(rel.name())] = HornRelation(rel)
+
+    def __str__(self):
+        out = io.StringIO()
+        for r in self._rules:
+            out.write(str(r))
+            out.write('\n')
+        out.write('\n')
+        for q in self._queries:
+            out.write(str(q))
+        return out.getvalue()
+
+    def load_from_fp(self, fp, queries):
+        self._fp = fp
+        if len(queries) > 0:
+            for r in fp.get_rules():
+                rule = HornRule(r)
+                #import pdb; pdb.set_trace()
+                self.add_rule(rule)
+            for q in queries:
+                rule = HornRule(z3.Implies(q, z3.BoolVal(False)))
+                self.add_rule(rule)
+        else:
+            # fixedpoit object is not properly loaded, ignore it
+            self._fp = None
+            for a in fp.get_assertions():
+                #import pdb; pdb.set_trace()
+                rule = HornRule(a) #doesn't work?
+                self.add_rule(rule)
+        self.seal()
+
+    def has_fixedpoint(self):
+        return self._fp is not None
+    def get_fixedpoint(self):
+        return self._fp
+
+    def mk_fixedpoint(self, fp = None):
+        if fp is None:
+            self._fp = z3.Fixedpoint()
+            fp = self._fp
+
+        for rel in self._rels_set:
+            fp.register_relation(rel) ##SIMILAR thing!
+        for r in self._rules:
+            if r.has_formula():
+                fp.add_rule(r.get_formula())
+            else:
+                fp.add_rule(r.mk_formula())
+
+        return fp
+
+
+class FolModel(object):
+    def __init__(self):
+        self._fn_interps = dict()
+
+    def add_fn(self, name, lmbd):
+        self._fn_interps[name] = lmbd
+
+    def has_interp(self, name):
+        return name in self._fn_interps.keys()
+
+    def __setitem__(self, key, val):
+        self.add_fn(key, val)
+
+    def get_fn(self, name):
+        return self._fn_interps[name]
+
+    def eval(self, term):
+        fn = self.get_fn(term.decl().name())
+        # lambdas seem to be broken at the moment
+        # this is a work around
+        body = fn.body()
+        body = z3.substitute_vars(body, *reversed(term.children()))
+        return body
+
+
+    def __str__(self):
+        return str(self._fn_interps)
+    def __repr__(self):
+        return repr(self._fn_interps)
+
+def load_horn_db_from_file(fname):
+    fp = z3.Fixedpoint()
+    queries = fp.parse_file(fname)
+    db = HornClauseDb(fname)
+    db.load_from_fp(fp, queries)
+    #import pdb; pdb.set_trace()
+    return db
+
+def copy_horn_db_mk_pinit(origdb, newParList=[z3.Const('U', z3.RealSort())], pInitPre=z3.BoolVal(True)):
```

Comment:
comment on what this function is expected to do would be nice

Reply:
Takes a linear CHC transition system benchmark (in SMTLib format), where CHCs expressible as 3 Horn Xlauses.

Copy the "HornDatabase" and mutate CHCs to:
   1. swap first constant into parameter U, and
   2. create "PInit" version of CHC system, by inserting new clauses/rules.